<hr style="border: 1px solid gray;">

# <code style="background:yellow;color:black">Summary of today's session:</code> 

1. Revision of last session.
2. loc accessor.
3. loc accessor with & and | symbols and "isin" keyword.
4. Application of any custom user-defined function on columns using apply() method.
5. Grouping and groupby concept and groupby() method in pandas. 
6. Groupby object and its methods like get_group() and attributes like groups and ngroups.
7. Multiple aggregations.
8. Groupby filtering and filter method.
9. A concept similar to window functions concept in sql.

# <code style="background:yellow;color:black">Revision of last lecture:</code>

**In last lecture, we introduced the imdb dataset, so, today we will be working on the same dataset.**

In [1]:
import numpy as np
import pandas as pd
import matplotlib as plt
import seaborn as sns

**Downloading imdb dataset from google drive link.**

In [2]:
!gdown 1s2TkjSpzNc4SyxqRrQleZyDIHlc7bxnd
!gdown 1Ws-_s1fHZ9nHfGLVUQurbHDvStePlEJm

Downloading...
From: https://drive.google.com/uc?id=1s2TkjSpzNc4SyxqRrQleZyDIHlc7bxnd
To: C:\Users\admin\OneDrive\Desktop\Scaler\Module 5- Python Libraries\My Practice\Class Practice\movies.csv

  0%|          | 0.00/112k [00:00<?, ?B/s]
100%|##########| 112k/112k [00:00<00:00, 721kB/s]
100%|##########| 112k/112k [00:00<00:00, 721kB/s]
Downloading...
From: https://drive.google.com/uc?id=1Ws-_s1fHZ9nHfGLVUQurbHDvStePlEJm
To: C:\Users\admin\OneDrive\Desktop\Scaler\Module 5- Python Libraries\My Practice\Class Practice\directors.csv

  0%|          | 0.00/65.4k [00:00<?, ?B/s]
100%|##########| 65.4k/65.4k [00:00<00:00, 661kB/s]


**Loading the dataset in pandas df.**

In [2]:
movies = pd.read_csv("movies.csv")
directors = pd.read_csv("directors.csv")

**Checking the dataset.**

In [4]:
movies.head()

Unnamed: 0     id     budget  popularity     revenue  \
0           0  43597  237000000         150  2787965087   
1           1  43598  300000000         139   961000000   
2           2  43599  245000000         107   880674609   
3           3  43600  250000000         112  1084939099   
4           5  43602  258000000         115   890871626   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   
2                                   Spectre           6.3        4466   
3                     The Dark Knight Rises           7.6        9106   
4                              Spider-Man 3           5.9        3576   

   director_id  year month       day  
0         4762  2009   Dec  Thursday  
1         4763  2007   May  Saturday  
2         4764  2015   Oct    Monday  
3         4765  2012   Jul    Monday  
4         4767  2007   May   Tuesday

**Removing the unnecessary column called unnamed: 0 in movies dataset.**

In [3]:
movies.drop(columns = ["Unnamed: 0"] , inplace = True)

**Removing the unnecessary column called unnamed: 0 in directors dataset.**

In [4]:
directors.drop(columns = ["Unnamed: 0"] , inplace = True)

**Checking both the datasets**

In [5]:
movies.head()

id     budget  popularity     revenue  \
0  43597  237000000         150  2787965087   
1  43598  300000000         139   961000000   
2  43599  245000000         107   880674609   
3  43600  250000000         112  1084939099   
4  43602  258000000         115   890871626   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   
2                                   Spectre           6.3        4466   
3                     The Dark Knight Rises           7.6        9106   
4                              Spider-Man 3           5.9        3576   

   director_id  year month       day  
0         4762  2009   Dec  Thursday  
1         4763  2007   May  Saturday  
2         4764  2015   Oct    Monday  
3         4765  2012   Jul    Monday  
4         4767  2007   May   Tuesday

In [6]:
directors.head()

director_name    id gender
0      James Cameron  4762   Male
1     Gore Verbinski  4763   Male
2         Sam Mendes  4764   Male
3  Christopher Nolan  4765   Male
4     Andrew Stanton  4766   Male

* **<code style="background:yellow;color:black">We want to have a single table, where we will have movie information and its director information attached to it.</code> So what concept can help us? -- MERGE or JOIN. .merge() method helps us in merging dataframes.** 
* Left join because we want all the movies information.
* Common column is in movies (left) table is "director_id" and in directors (right) table is "id".
* We are storing the new merged table in a new df called "data".
* **<code style="background:yellow;color:black">An important observation</code> is that movie's id column became "id_x" and director's id column became "id_y" automatically. Because both the tables had id columns so they were renamed as id_x and id_y to distinguish between both.** 

In [7]:
data = movies.merge(directors, how = "left", left_on = "director_id", right_on = "id")

data.head()

id_x     budget  popularity     revenue  \
0  43597  237000000         150  2787965087   
1  43598  300000000         139   961000000   
2  43599  245000000         107   880674609   
3  43600  250000000         112  1084939099   
4  43602  258000000         115   890871626   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   
2                                   Spectre           6.3        4466   
3                     The Dark Knight Rises           7.6        9106   
4                              Spider-Man 3           5.9        3576   

   director_id  year month       day      director_name  id_y gender  
0         4762  2009   Dec  Thursday      James Cameron  4762   Male  
1         4763  2007   May  Saturday     Gore Verbinski  4763   Male  
2         4764  2015   Oct    Monday         Sam Mendes  4764   Male  
3         4765  2012   Jul    Monday  Christopher Nolan  4765   Male  
4         4767  2007   May   Tuesday          Sam Raimi  4767   Male

**We dont need "id_y" because we already have director_id column, so we will drop it.** 

In [8]:
data.drop(columns = ["id_y"], inplace = True)

data.head()

id_x     budget  popularity     revenue  \
0  43597  237000000         150  2787965087   
1  43598  300000000         139   961000000   
2  43599  245000000         107   880674609   
3  43600  250000000         112  1084939099   
4  43602  258000000         115   890871626   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   
2                                   Spectre           6.3        4466   
3                     The Dark Knight Rises           7.6        9106   
4                              Spider-Man 3           5.9        3576   

   director_id  year month       day      director_name gender  
0         4762  2009   Dec  Thursday      James Cameron   Male  
1         4763  2007   May  Saturday     Gore Verbinski   Male  
2         4764  2015   Oct    Monday         Sam Mendes   Male  
3         4765  2012   Jul    Monday  Christopher Nolan   Male  
4         4767  2007   May   Tuesday          Sam Raimi   Male

* **We can see that in total we have 1465 values, but in gender column we have 1341 non null values, so we have some null values in gender column.** 
* How to deal with missing data, we will learn in next class.

In [11]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1465 entries, 0 to 1464
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id_x           1465 non-null   int64  
 1   budget         1465 non-null   int64  
 2   popularity     1465 non-null   int64  
 3   revenue        1465 non-null   int64  
 4   title          1465 non-null   object 
 5   vote_average   1465 non-null   float64
 6   vote_count     1465 non-null   int64  
 7   director_id    1465 non-null   int64  
 8   year           1465 non-null   int64  
 9   month          1465 non-null   object 
 10  day            1465 non-null   object 
 11  director_name  1465 non-null   object 
 12  gender         1341 non-null   object 
dtypes: float64(1), int64(7), object(5)
memory usage: 160.2+ KB


**In last lecture, we learnt this method which gives us statistical information about our dataset.**

In [12]:
data.describe()

id_x        budget   popularity       revenue  vote_average  \
count   1465.000000  1.465000e+03  1465.000000  1.465000e+03   1465.000000   
mean   45225.191126  4.802295e+07    30.855973  1.432539e+08      6.368191   
std     1189.096396  4.935541e+07    34.845214  2.064918e+08      0.818033   
min    43597.000000  0.000000e+00     0.000000  0.000000e+00      3.000000   
25%    44236.000000  1.400000e+07    11.000000  1.738013e+07      5.900000   
50%    45022.000000  3.300000e+07    23.000000  7.578164e+07      6.400000   
75%    45990.000000  6.600000e+07    41.000000  1.792469e+08      6.900000   
max    48395.000000  3.800000e+08   724.000000  2.787965e+09      8.300000   

         vote_count  director_id         year  
count   1465.000000  1465.000000  1465.000000  
mean    1146.396587  5040.192491  2002.615017  
std     1578.077438   258.059631     8.680141  
min        1.000000  4762.000000  1976.000000  
25%      216.000000  4845.000000  1998.000000  
50%      571.000000  4964.000000  2004.000000  
75%     1387.000000  5179.000000  2009.000000  
max    13752.000000  6204.000000  2016.000000

In [13]:
data.describe(include = "object")

title month     day     director_name gender
count     1465  1465    1465              1465   1341
unique    1465    12       7               199      2
top     Avatar   Dec  Friday  Steven Spielberg   Male
freq         1   193     654                26   1309

* **We had a homework about converting numbers into millions of revenue as well like we did of budget.**
* **We can even round the data in millions to two decimal places with the use of <u>.round() method</u> for both budget and revenue.**

In [9]:
data["revenue"] = (data["revenue"]/1000000).round(2)
data["budget"] = (data["budget"]/1000000).round(2)

data.head()

id_x  budget  popularity  revenue  \
0  43597   237.0         150  2787.97   
1  43598   300.0         139   961.00   
2  43599   245.0         107   880.67   
3  43600   250.0         112  1084.94   
4  43602   258.0         115   890.87   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   
2                                   Spectre           6.3        4466   
3                     The Dark Knight Rises           7.6        9106   
4                              Spider-Man 3           5.9        3576   

   director_id  year month       day      director_name gender  
0         4762  2009   Dec  Thursday      James Cameron   Male  
1         4763  2007   May  Saturday     Gore Verbinski   Male  
2         4764  2015   Oct    Monday         Sam Mendes   Male  
3         4765  2012   Jul    Monday  Christopher Nolan   Male  
4         4767  2007   May   Tuesday          Sam Raimi   Male

<hr style="border: 1px solid gray;">

### <code style="background:yellow;color:black">loc accessor:</code>

* **In pandas, the <code style="background:yellow;color:black">.loc accessor</code> is used for label-based indexing. It is primarily used to select rows and columns from a DataFrame by their labels or a boolean array.**

* **The basic syntax for using .loc is: <code style="background:yellow;color:black">DataFrame.loc[row_indexer, column_indexer]</code>**

**<u>Here, row_indexer and column_indexer can take various forms:**</u>

* **<code style="background:yellow;color:black">Label-based selection:</code>** If we provide a single label for row_indexer, it selects the corresponding row. If we provide a single label for column_indexer, it selects the corresponding column. If we provide labels for both row and column indexers, it selects the intersection of the specified row and column. **Example:** Selecting a single row and multiple columns - **df.loc['row_label', ['column1', 'column2']]**

* **<code style="background:yellow;color:black">Slicing with labels:</code>** We can use label-based slicing to select a range of rows or columns. **Example:** Selecting a range of rows and columns - **df.loc['row_label1':'row_label2', 'column_label1':'column_label2']**

* **<code style="background:yellow;color:black">Boolean indexing:</code>** We can use boolean conditions to filter rows based on a certain criterion. **Example:** Selecting rows based on a condition - **df.loc[df['column'] > 5]**

* **<code style="background:yellow;color:black">Selecting all rows for specific columns:</code>** We can use "**:**" as a wildcard to select all rows for specific columns. **Example:** Selecting all rows for specific columns - **df.loc[:, ['column1', 'column2']]**

We have to keep in mind that .loc is inclusive on both ends when using label-based slicing, unlike integer-based slicing, which is exclusive on the upper bound.

<u>Here's a simple example:</u>

In [15]:
# Creating a sample DataFrame
data1 = {
    'A': [1, 2, 3],
    'B': [4, 5, 6],
    'C': [7, 8, 9]
}

df1 = pd.DataFrame(data1, index = ['X', 'Y', 'Z'])

# Using .loc to select specific rows and columns
selected_data = df1.loc[['X', 'Y'], ['A', 'B']]
selected_data

A  B
X  1  4
Y  2  5

<u><code style="background:yellow;color:black">Here's another simple example:</u></code>

* **For the below mask, we will get boolean true false values for the given condition for all the "vote_average" column rows.**
* **We want to filter out only those rows where vote_average > 7.** 
* **We can implement that using the code below. We will use loc and provide the mask within it.** 
* This gives us 301 rows (or movies) where condition is met (or where vote_average > 7). 
* We are storing these filtered rows or data in a df2 dataframe.
* Shape of df2 will be 301 X 13 as can be seen on output of data.loc[data["vote_average"] > 7] --> 301 rows × 13 columns.

In [16]:
data["vote_average"] > 7

0        True
1       False
2       False
3        True
4       False
        ...  
1460     True
1461     True
1462    False
1463    False
1464    False
Name: vote_average, Length: 1465, dtype: bool

In [17]:
data.loc[data["vote_average"] > 7]

id_x  budget  popularity  revenue  \
0     43597  237.00         150  2787.97   
3     43600  250.00         112  1084.94   
14    43616  250.00         120   956.02   
16    43619  250.00          94   958.40   
19    43622  200.00         100  1845.03   
...     ...     ...         ...      ...   
1456  48321    0.01          20     7.00   
1457  48323    0.00           5     0.00   
1458  48335    0.06          27     3.22   
1460  48363    0.00           3     0.32   
1461  48370    0.03          19     3.15   

                                          title  vote_average  vote_count  \
0                                        Avatar           7.2       11800   
3                         The Dark Knight Rises           7.6        9106   
14    The Hobbit: The Battle of the Five Armies           7.1        4760   
16          The Hobbit: The Desolation of Smaug           7.6        4524   
19                                      Titanic           7.5        7562   
...                                         ...           ...         ...   
1456                                 Eraserhead           7.5         485   
1457                                 The Mighty           7.1          51   
1458                                         Pi           7.1         586   
1460                             The Last Waltz           7.9          64   
1461                                     Clerks           7.4         755   

      director_id  year month        day      director_name gender  
0            4762  2009   Dec   Thursday      James Cameron   Male  
3            4765  2012   Jul     Monday  Christopher Nolan   Male  
14           4777  2014   Dec  Wednesday      Peter Jackson   Male  
16           4777  2013   Dec  Wednesday      Peter Jackson   Male  
19           4762  1997   Nov    Tuesday      James Cameron   Male  
...           ...   ...   ...        ...                ...    ...  
1456         5265  1977   Mar   Saturday        David Lynch   Male  
1457         4921  1998   Oct     Friday      Peter Chelsom   Male  
1458         4881  1998   Jul     Friday   Darren Aronofsky   Male  
1460         4809  1978   May     Monday    Martin Scorsese   Male  
1461         5369  1994   Sep    Tuesday        Kevin Smith   Male  

[301 rows x 13 columns]

In [18]:
df2 = data.loc[data["vote_average"] > 7]

df2.head()

id_x  budget  popularity  revenue  \
0   43597   237.0         150  2787.97   
3   43600   250.0         112  1084.94   
14  43616   250.0         120   956.02   
16  43619   250.0          94   958.40   
19  43622   200.0         100  1845.03   

                                        title  vote_average  vote_count  \
0                                      Avatar           7.2       11800   
3                       The Dark Knight Rises           7.6        9106   
14  The Hobbit: The Battle of the Five Armies           7.1        4760   
16        The Hobbit: The Desolation of Smaug           7.6        4524   
19                                    Titanic           7.5        7562   

    director_id  year month        day      director_name gender  
0          4762  2009   Dec   Thursday      James Cameron   Male  
3          4765  2012   Jul     Monday  Christopher Nolan   Male  
14         4777  2014   Dec  Wednesday      Peter Jackson   Male  
16         4777  2013   Dec  Wednesday      Peter Jackson   Male  
19         4762  1997   Nov    Tuesday      James Cameron   Male

In [19]:
df2.shape

(301, 13)

<hr style="border: 1px solid gray;">

**<code style="background:yellow;color:black">Question 1:</code> Display the title and director_name of the movies with vote_avg > 7.**

* **We can do slicing and multi indexing because from df2, we want specifically 2 columns i.e. title and director_name.**
* **We want all the rows so we are doing ':', and then column names.**
* If we want budget also here, we can add budget column name as well.
* **This code cell below is in a fancy indexing already because we are using df2 which is already masked data instead of "data" dataset.**
* **If we want to write the above code in single way or using <u>"data"</u> dataset which isn't masked, what we can write is like this.**
* **In next code cell below, in data.loc, in rows area, we are providing the condition i.e. data["vote_average"] > 7 and in columns area we are providing column names.** 

**<code style="background:yellow;color:black">Format: data.loc[row_condition , column_condition]</code>**

In [20]:
df2.loc[: , ["title" , "director_name"]]

title      director_name
0                                        Avatar      James Cameron
3                         The Dark Knight Rises  Christopher Nolan
14    The Hobbit: The Battle of the Five Armies      Peter Jackson
16          The Hobbit: The Desolation of Smaug      Peter Jackson
19                                      Titanic      James Cameron
...                                         ...                ...
1456                                 Eraserhead        David Lynch
1457                                 The Mighty      Peter Chelsom
1458                                         Pi   Darren Aronofsky
1460                             The Last Waltz    Martin Scorsese
1461                                     Clerks        Kevin Smith

[301 rows x 2 columns]

In [21]:
data.loc[ data["vote_average"] > 7 , ["title", "director_name"]]

title      director_name
0                                        Avatar      James Cameron
3                         The Dark Knight Rises  Christopher Nolan
14    The Hobbit: The Battle of the Five Armies      Peter Jackson
16          The Hobbit: The Desolation of Smaug      Peter Jackson
19                                      Titanic      James Cameron
...                                         ...                ...
1456                                 Eraserhead        David Lynch
1457                                 The Mighty      Peter Chelsom
1458                                         Pi   Darren Aronofsky
1460                             The Last Waltz    Martin Scorsese
1461                                     Clerks        Kevin Smith

[301 rows x 2 columns]

<hr style="border: 1px solid gray;">

**<code style="background:yellow;color:black">Question 2:</code> Display only those movies where the vote_avg > 7 and they have been released in or after 2013.**

* How many filtering criteria will be required? -- TWO.
* And those 2 filtering criteria will have AND or OR in between them? -- AND.
* **Whenever we are writing multiple conditions in the filtering, we must always write each condition in bracket.** 
* **When we run the below code, we get an error because in pandas when we are writing AND or OR, we have to write logical AND OR notation.**
* **So, below to below code works because it has "&"**
* These are all the movies which have vote_average > 7 and which have been released on or after 2013.
<br> <br> 
* **<code style="background:yellow;color:black">and --> &**</code> 
* **<code style="background:yellow;color:black">or --> |**</code>

In [22]:
data.loc[(data["vote_average"] > 7) and (data["year"] >= 2013)]

ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().

In [23]:
data.loc[(data["vote_average"] > 7) & (data["year"] >= 2013)]

id_x  budget  popularity  revenue  \
14    43616   250.0         120   956.02   
16    43619   250.0          94   958.40   
30    43641   190.0         102  1506.25   
32    43643   250.0         118   747.86   
52    43671   178.0          79   370.54   
58    43692   165.0         724   675.12   
78    43724   150.0         434   378.86   
106   43773   135.0         100   532.95   
109   43780   130.0          76   847.42   
162   43867   108.0         167   630.16   
183   43895   100.0          95   392.00   
273   44053    68.0         139   211.82   
302   44116    75.0          71   117.70   
312   44128    75.0          48   108.15   
394   44281    44.0          68   155.76   
398   44290    61.0         143   369.33   
447   44387    58.8          87   542.31   
465   44415    55.0          52    95.00   
467   44418    55.0          79   192.33   
616   44755    40.0          37   149.30   
625   44770    35.0          53   194.56   
635   44784    40.0          48   165.48   
692   44929    38.0          33    90.25   
776   45129    30.0          74   174.60   
808   45194    30.0          65    91.71   
833   45293    28.0          61   201.63   
839   45301    28.0          57   133.35   
975   45622    22.0          27    88.88   
1002  45693    13.0          49   318.00   
1077  45931    18.0          89   103.22   
1210  46440    12.0          29   100.13   
1211  46446    12.0          29    17.65   
1227  46512     0.0          14     0.00   
1238  46548    11.0          36    32.94   
1344  47181     5.0          22    24.80   
1372  47350     4.0          43    44.35   
1391  47515     3.0          23    11.18   

                                          title  vote_average  vote_count  \
14    The Hobbit: The Battle of the Five Armies           7.1        4760   
16          The Hobbit: The Desolation of Smaug           7.6        4524   
30                                    Furious 7           7.3        4176   
32                   X-Men: Days of Future Past           7.5        6032   
52                             Edge of Tomorrow           7.6        4858   
58                                 Interstellar           8.1       10867   
78                           Mad Max: Fury Road           7.2        9427   
106                                The Revenant           7.3        6396   
109             The Hunger Games: Catching Fire           7.4        6495   
162                                 The Martian           7.6        7268   
183                     The Wolf of Wall Street           7.9        6571   
273                                        Fury           7.4        3946   
302                              Now You See Me           7.3        5487   
312                     The Man from U.N.C.L.E.           7.1        2265   
394                           The Hateful Eight           7.6        4274   
398                                   Gone Girl           7.9        5862   
447                             American Sniper           7.4        4469   
465                            Captain Phillips           7.6        2454   
467                               The Equalizer           7.1        2954   
616                               Lone Survivor           7.4        1526   
625                                  The Intern           7.1        1881   
635                             Bridge of Spies           7.2        2583   
692                                        Rush           7.7        2277   
776                    The Grand Budapest Hotel           8.0        4519   
808                                    Southpaw           7.3        2067   
833                      Straight Outta Compton           7.7        1355   
839                               The Big Short           7.3        2607   
975                    The Hundred-Foot Journey           7.3         508   
1002                              The Conjuring           7.4        3092   
1077                    

<hr style="border: 1px solid gray;">

**<code style="background:yellow;color:black">Question 3:</code> Display those rows where movie has been released on Friday or Saturday.**

* In below code cell, in data.loc we have put both conditions of Friday or Saturday. 
* Those rows where the conditions are met are filtered out.
* **In SQL, we have a keyword called "in" which we used when we have multiple OR conditions.** 
* **In pandas we have .isin(). So, we can have the following code using .isin() like this below. Result is same as compared to previous result.**
* This is more sophisticated way of doing the same.

In [24]:
data.head()

id_x  budget  popularity  revenue  \
0  43597   237.0         150  2787.97   
1  43598   300.0         139   961.00   
2  43599   245.0         107   880.67   
3  43600   250.0         112  1084.94   
4  43602   258.0         115   890.87   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   
2                                   Spectre           6.3        4466   
3                     The Dark Knight Rises           7.6        9106   
4                              Spider-Man 3           5.9        3576   

   director_id  year month       day      director_name gender  
0         4762  2009   Dec  Thursday      James Cameron   Male  
1         4763  2007   May  Saturday     Gore Verbinski   Male  
2         4764  2015   Oct    Monday         Sam Mendes   Male  
3         4765  2012   Jul    Monday  Christopher Nolan   Male  
4         4767  2007   May   Tuesday          Sam Raimi   Male

In [25]:
data.loc[(data["day"] == "Saturday") | (data["day"] == "Friday")]

id_x  budget  popularity  revenue  \
1     43598  300.00         139   961.00   
12    43614  380.00         135  1045.71   
22    43627  200.00          35   783.77   
25    43632  150.00          21   836.30   
40    43656  200.00          45   769.65   
...     ...     ...         ...      ...   
1457  48323    0.00           5     0.00   
1458  48335    0.06          27     3.22   
1462  48375    0.00           7     0.00   
1463  48376    0.00           3     0.00   
1464  48395    0.22          14     2.04   

                                            title  vote_average  vote_count  \
1        Pirates of the Caribbean: At World's End           6.9        4500   
12    Pirates of the Caribbean: On Stranger Tides           6.4        4948   
22                                   Spider-Man 2           6.7        4321   
25            Transformers: Revenge of the Fallen           6.0        3138   
40                                           2012           5.6        4903   
...                                           ...           ...         ...   
1457                                   The Mighty           7.1          51   
1458                                           Pi           7.1         586   
1462                                      Rampage           6.0         131   
1463                                      Slacker           6.4          77   
1464                                  El Mariachi           6.6         238   

      director_id  year month       day      director_name gender  
1            4763  2007   May  Saturday     Gore Verbinski   Male  
12           4775  2011   May  Saturday       Rob Marshall   Male  
22           4767  2004   Jun    Friday          Sam Raimi   Male  
25           4788  2009   Jun    Friday        Michael Bay   Male  
40           4803  2009   Oct  Saturday    Roland Emmerich   Male  
...           ...   ...   ...       ...                ...    ...  
1457         4921  1998   Oct    Friday      Peter Chelsom   Male  
1458         4881  1998   Jul    Friday   Darren Aronofsky   Male  
1462         5148  2009   Aug    Friday           Uwe Boll   Male  
1463         5535  1990   Jul    Friday  Richard Linklater   Male  
1464         5097  1992   Sep    Friday   Robert Rodriguez    NaN  

[701 rows x 13 columns]

In [26]:
data.loc[data["day"].isin(["Saturday", "Friday"])]

id_x  budget  popularity  revenue  \
1     43598  300.00         139   961.00   
12    43614  380.00         135  1045.71   
22    43627  200.00          35   783.77   
25    43632  150.00          21   836.30   
40    43656  200.00          45   769.65   
...     ...     ...         ...      ...   
1457  48323    0.00           5     0.00   
1458  48335    0.06          27     3.22   
1462  48375    0.00           7     0.00   
1463  48376    0.00           3     0.00   
1464  48395    0.22          14     2.04   

                                            title  vote_average  vote_count  \
1        Pirates of the Caribbean: At World's End           6.9        4500   
12    Pirates of the Caribbean: On Stranger Tides           6.4        4948   
22                                   Spider-Man 2           6.7        4321   
25            Transformers: Revenge of the Fallen           6.0        3138   
40                                           2012           5.6        4903   
...                                           ...           ...         ...   
1457                                   The Mighty           7.1          51   
1458                                           Pi           7.1         586   
1462                                      Rampage           6.0         131   
1463                                      Slacker           6.4          77   
1464                                  El Mariachi           6.6         238   

      director_id  year month       day      director_name gender  
1            4763  2007   May  Saturday     Gore Verbinski   Male  
12           4775  2011   May  Saturday       Rob Marshall   Male  
22           4767  2004   Jun    Friday          Sam Raimi   Male  
25           4788  2009   Jun    Friday        Michael Bay   Male  
40           4803  2009   Oct  Saturday    Roland Emmerich   Male  
...           ...   ...   ...       ...                ...    ...  
1457         4921  1998   Oct    Friday      Peter Chelsom   Male  
1458         4881  1998   Jul    Friday   Darren Aronofsky   Male  
1462         5148  2009   Aug    Friday           Uwe Boll   Male  
1463         5535  1990   Jul    Friday  Richard Linklater   Male  
1464         5097  1992   Sep    Friday   Robert Rodriguez    NaN  

[701 rows x 13 columns]

<hr style="border: 1px solid gray;">

**<code style="background:yellow;color:black">Question 4:</code> How to filter those records or rows where movies were released in the year(2015, 2016, 2012) from the above dataset?**

In [27]:
data["year"].isin([2015, 2016, 2012])

0       False
1       False
2        True
3        True
4       False
        ...  
1460    False
1461    False
1462    False
1463    False
1464    False
Name: year, Length: 1465, dtype: bool

<hr style="border: 1px solid gray;">

# <code style="background:yellow;color:black">apply() method:</code>

* **In pandas, the <code style="background:yellow;color:black">apply() method</code> is used to apply a function along the axis of a DataFrame or Series. The apply method is particularly useful when we want to perform a custom operation on the elements of a DataFrame or Series.**
* **To understand .apply(), lets do this task:** 
    * Task: Give encoding to the Genders: 
        * Male --> 0 
        * Female --> 1 

    * Lets write a function for this, which we have implemented in the below code cell. <br><br>

* **<code style="background:yellow;color:black">Now, we want to apply this function on the entire column.</code> So, we apply here .apply() method here in next code cells.** 
* This is similar to vectorization. <br><br>

* **<u><code style="background:yellow;color:black">Here's a step-by-step breakdown:</code></u>**

    1. The encode_gender function is defined, which takes a parameter "g" (presumably representing gender) and returns 0 if "g" is equal to "Male" and 1 otherwise.

    2. The apply method is used on the "gender" column of the DataFrame "data". For each value in the "gender" column, the encode_gender function is applied, resulting in the transformation of "Male" to 0 and any other value to 1.

    3. The transformed values are then assigned back to the "gender" column in the DataFrame data. <br><br>

* **So, after executing this code, the "gender" column in the DataFrame data will have values replaced with 0 for "Male" and 1 for any other gender.**

* **It will apply row by row that function and it will store the result back into gender column. We can create another column as well but, here we are modifying the same column.**

* In gender column, we have got encoding. All males have been given 0 and females 1.

* **<code style="background:yellow;color:black">This is how we can apply a particular function on a particular column.</code>**

**This kind of encoding is useful when we want to convert categorical variables into a format that can be used in machine learning models, especially models that require numerical inputs.**

In [10]:
def encode_gender(g):
    
    if g == "Male":
        return 0
    
    else:
        return 1

In [11]:
encode_gender("Male")

0

In [12]:
data["gender"] = data["gender"].apply(encode_gender)

data.head()

id_x  budget  popularity  revenue  \
0  43597   237.0         150  2787.97   
1  43598   300.0         139   961.00   
2  43599   245.0         107   880.67   
3  43600   250.0         112  1084.94   
4  43602   258.0         115   890.87   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   
2                                   Spectre           6.3        4466   
3                     The Dark Knight Rises           7.6        9106   
4                              Spider-Man 3           5.9        3576   

   director_id  year month       day      director_name  gender  
0         4762  2009   Dec  Thursday      James Cameron       0  
1         4763  2007   May  Saturday     Gore Verbinski       0  
2         4764  2015   Oct    Monday         Sam Mendes       0  
3         4765  2012   Jul    Monday  Christopher Nolan       0  
4         4767  2007   May   Tuesday          Sam Raimi       0

In [31]:
# Creating a sample DataFrame
data3 = {
    'A': [1, 2, 3],
    'B': [4, 5, 6],
    'C': [7, 8, 9]
}

df3 = pd.DataFrame(data3)

# Define a custom function
def custom_function(x):
    return x * 2

# Applying the custom function to each element of the DataFrame
result = df3.apply(custom_function)
result

A   B   C
0  2   8  14
1  4  10  16
2  6  12  18

* **Here we need to understand 2 things: by default axis = 0 (vertical axis), vertical axis means column wise.**
* **np.sum is also a function. Now by default axis is 0, and we are doing np.sum on 2 columns, so how many values will be the there? -- 2 values.**
* It will perform the entire sum function over the revenue column and budget column as well. 
* **Here's a breakdown of the code:**
    1. data[["revenue", "budget"]]: This part of the code selects a subset of the DataFrame data containing only the "revenue" and "budget" columns.
    2. .apply(np.sum): The apply method is then used to apply the np.sum function to each column in the selected subset. This results in the sum of each column.
* So, after executing this code, we will get a Pandas Series with the sum of "revenue" and "budget" columns. The index of the resulting Series will likely contain the column names ("revenue" and "budget"), and the corresponding values will be the sums of each respective column.

In [13]:
data[["revenue", "budget"]].apply(np.sum) 

revenue    209867.04
budget      70353.62
dtype: float64

* **If we change the above code to axis = 1 (row by row), so how many values will be the there? -- 1465 values.**
* So, for every row, we will get sum of revenue and budget as total, and we will thus get 1465 values. 
* **Here's a breakdown:**
    1. data[["revenue", "budget"]]: This selects a subset of the DataFrame data containing only the "revenue" and "budget" columns.
    2. .apply(np.sum, axis=1): The apply method is used to apply the np.sum function along axis = 1, meaning it performs the sum operation for each row. The result is a Pandas Series where each element represents the sum of "revenue" and "budget" for the corresponding row.

In [14]:
data[["revenue", "budget"]].apply(np.sum, axis = 1)

0       3024.97
1       1261.00
2       1125.67
3       1334.94
4       1148.87
         ...   
1460       0.32
1461       3.18
1462       0.00
1463       0.00
1464       2.26
Length: 1465, dtype: float64

In [34]:
data.shape

(1465, 13)

In [35]:
data.head()

id_x  budget  popularity  revenue  \
0  43597   237.0         150  2787.97   
1  43598   300.0         139   961.00   
2  43599   245.0         107   880.67   
3  43600   250.0         112  1084.94   
4  43602   258.0         115   890.87   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   
2                                   Spectre           6.3        4466   
3                     The Dark Knight Rises           7.6        9106   
4                              Spider-Man 3           5.9        3576   

   director_id  year month       day      director_name  gender  
0         4762  2009   Dec  Thursday      James Cameron       0  
1         4763  2007   May  Saturday     Gore Verbinski       0  
2         4764  2015   Oct    Monday         Sam Mendes       0  
3         4765  2012   Jul    Monday  Christopher Nolan       0  
4         4767  2007   May   Tuesday          Sam Raimi       0

**Creating a new function to calculate profit.**

In [15]:
def calc_profit(df):
    
    profit = df["revenue"] - df["budget"]
    
    return profit

**Creating a new colum called profit. Passing 2 colums of revenue and budget in the calc_profit function, axis = 1 because we want row wise calculation of profit.**

In [16]:
data["profit"] = data[["revenue", "budget"]].apply(calc_profit, axis = 1)

**Here, in new colum profit, we can see profit of each row.**

In [17]:
data.head()

id_x  budget  popularity  revenue  \
0  43597   237.0         150  2787.97   
1  43598   300.0         139   961.00   
2  43599   245.0         107   880.67   
3  43600   250.0         112  1084.94   
4  43602   258.0         115   890.87   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   
2                                   Spectre           6.3        4466   
3                     The Dark Knight Rises           7.6        9106   
4                              Spider-Man 3           5.9        3576   

   director_id  year month       day      director_name  gender   profit  
0         4762  2009   Dec  Thursday      James Cameron       0  2550.97  
1         4763  2007   May  Saturday     Gore Verbinski       0   661.00  
2         4764  2015   Oct    Monday         Sam Mendes       0   635.67  
3         4765  2012   Jul    Monday  Christopher Nolan       0   834.94  
4         4767  2007   May   Tuesday          Sam Raimi       0   632.87

**Just for clarity, following is a dataframe with 2 columns.**

In [18]:
data[["revenue", "budget"]]

revenue  budget
0     2787.97  237.00
1      961.00  300.00
2      880.67  245.00
3     1084.94  250.00
4      890.87  258.00
...       ...     ...
1460     0.32    0.00
1461     3.15    0.03
1462     0.00    0.00
1463     0.00    0.00
1464     2.04    0.22

[1465 rows x 2 columns]

* **We could have done this profit calc in one other way. Like the code below-** 
* This is also 100% correct way. And we can check profit and new_profit columns are exactly the same.
* **What we want here is that we get clear on apply() method, because in industry this is mostly used.**

In [19]:
data["new_profit"] = data["revenue"] - data["budget"]

**Checking the dataset for changes.**

In [20]:
data.head()

id_x  budget  popularity  revenue  \
0  43597   237.0         150  2787.97   
1  43598   300.0         139   961.00   
2  43599   245.0         107   880.67   
3  43600   250.0         112  1084.94   
4  43602   258.0         115   890.87   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   
2                                   Spectre           6.3        4466   
3                     The Dark Knight Rises           7.6        9106   
4                              Spider-Man 3           5.9        3576   

   director_id  year month       day      director_name  gender   profit  \
0         4762  2009   Dec  Thursday      James Cameron       0  2550.97   
1         4763  2007   May  Saturday     Gore Verbinski       0   661.00   
2         4764  2015   Oct    Monday         Sam Mendes       0   635.67   
3         4765  2012   Jul    Monday  Christopher Nolan       0   834.94   
4         4767  2007   May   Tuesday          Sam Raimi       0   632.87   

   new_profit  
0     2550.97  
1      661.00  
2      635.67  
3      834.94  
4      632.87

<hr style="border: 1px solid gray;">

# <code style="background:yellow;color:black">What is Grouping ? Group by in context of pandas</code>

**It could simply be understood through the terms -- split, apply, combine**

<img src="https://learning.oreilly.com/api/v2/epubs/urn:orm:book:9781491912126/files/assets/pyds_03in01.png" height = 350/>

1. **<code style="background:yellow;color:black">Split:</code> Breaking up and grouping** a DataFrame depending on the value of the specified key.

2. **<code style="background:yellow;color:black">Apply:</code>** Computing **some function**, usually an **aggregate, transformation, or filtering**, within the individual groups.

3. **<code style="background:yellow;color:black">Combine:</code> Merge the results** of these operations into an output array.

> Note: All these steps are to understand the topic

* Lets say we have a dataframe like the one shown in image above. Grouping can be understood in terms of split, apply and combine. 

* Lets say we want to group on the basis of "key" column. So what pandas does internally is it will first split the data "key-wise", so all the rows of A together, all the rows of B together and all the rows of C together. 

* Now, it will apply some aggregation like sum function. So, we are summing it up, so for all the A values we have sum as 5, for B 7 and for C 9. 

* And then finally, it will merge or combine. That means merge the results of the aggregate functions into an output array. So, if we see input and output just by looking, we can see its groupby and aggregation, we group by key and we get sum of each group. Same as SQL. 

* The beauty here is these things are running in parallel and this makes it very very fast. 

* How to implement in code, we will see below.

# <code style="background:yellow;color:black">groupby() method:</code>

* **The <code style="background:yellow;color:black">groupby() method</code> in pandas is used to split a DataFrame into groups based on some criteria, apply a function to each group independently, and then combine the results back into a DataFrame. It is a powerful and flexible tool for data analysis and manipulation.**
* The groupby() method is a key tool for analyzing and summarizing data based on different criteria, and it is commonly used in combination with other pandas functions to perform complex data manipulations and analyses.

In [21]:
data.head()

id_x  budget  popularity  revenue  \
0  43597   237.0         150  2787.97   
1  43598   300.0         139   961.00   
2  43599   245.0         107   880.67   
3  43600   250.0         112  1084.94   
4  43602   258.0         115   890.87   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   
2                                   Spectre           6.3        4466   
3                     The Dark Knight Rises           7.6        9106   
4                              Spider-Man 3           5.9        3576   

   director_id  year month       day      director_name  gender   profit  \
0         4762  2009   Dec  Thursday      James Cameron       0  2550.97   
1         4763  2007   May  Saturday     Gore Verbinski       0   661.00   
2         4764  2015   Oct    Monday         Sam Mendes       0   635.67   
3         4765  2012   Jul    Monday  Christopher Nolan       0   834.94   
4         4767  2007   May   Tuesday          Sam Raimi       0   632.87   

   new_profit  
0     2550.97  
1      661.00  
2      635.67  
3      834.94  
4      632.87

**We want to apply groupby over director column. Simply the column on the basis of which we want to groupby. When we see the output of this, we see that it is a groupby object. Its not a dataframe. <code style="background:yellow;color:black">This groupby object provides us with multiple functionalities.</code>** 

In [22]:
data.groupby("director_name")

* **<code style="background:yellow;color:black">Over groupby object we are applying one of its attributes using dot operator.</code>**
* **The <code style="background:yellow;color:black">.ngroups attribute</code> in pandas is used to get the number of groups formed after applying the groupby() method on a DataFrame.** 
* When we use groupby() to group a DataFrame based on one or more columns, it creates distinct groups based on the unique values in those columns. 
* The .ngroups attribute returns the number of such groups. It tells us how many groups are there. 
* So in our example, it means there are 199 director names or groups. 
* Here, every director name will be one group.

In [23]:
data.groupby("director_name").ngroups

199

* **We are lets say more interested to know how the groups are stored. We dont have to worry about how pandas does it internally.**
* **The <code style="background:yellow;color:black">.groups attribute</code> is often used to inspect the composition of groups after applying the groupby() method.**
* It shows us all the groups. 
* The output of this attribute is a dictionary. 
* First example says dict key is director name and dict values is a list of all the indices where the director name is Adam McKay.

In [24]:
data.groupby("director_name").groups

{'Adam McKay': [176, 323, 366, 505, 839, 916], 'Adam Shankman': [265, 300, 350, 404, 458, 843, 999, 1231], 'Alejandro González Iñárritu': [106, 749, 1015, 1034, 1077, 1405], 'Alex Proyas': [95, 159, 514, 671, 873], 'Alexander Payne': [793, 1006, 1101, 1211, 1281], 'Andrew Adamson': [11, 43, 328, 501, 947], 'Andrew Niccol': [533, 603, 701, 722, 1439], 'Andrzej Bartkowiak': [349, 549, 754, 911, 924], 'Andy Fickman': [517, 681, 909, 926, 973, 1023], 'Andy Tennant': [314, 320, 464, 593, 676, 885], 'Ang Lee': [99, 134, 748, 840, 1089, 1110, 1132, 1184], 'Anne Fletcher': [610, 650, 736, 789, 1206], 'Antoine Fuqua': [310, 338, 424, 467, 576, 808, 818, 1105], 'Atom Egoyan': [946, 1128, 1164, 1194, 1347, 1416], 'Barry Levinson': [313, 319, 471, 594, 878, 898, 1013, 1037, 1082, 1143, 1185, 1345, 1378], 'Barry Sonnenfeld': [13, 48, 90, 205, 591, 778, 783], 'Ben Stiller': [209, 212, 547, 562, 850], 'Bill Condon': [102, 307, 902, 1233, 1381], 'Bobby Farrelly': [352, 356, 481, 498, 624, 630, 654, 80

* **So if we want to extract all the information of a particular group? For that we have a method called <code style="background:yellow;color:black">get_group().</code>**
* **The <code style="background:yellow;color:black">get_group() method</code> in pandas is used to retrieve a specific group from a grouped DataFrame. After using the groupby() method to create groups based on certain criteria, we can use get_group() to extract a particular group by specifying the key associated with that group.**
* It actually gives us that particular group's all rows. 
* Also note that all the indices were there in the output of .groups as well for Adam McKay. 

In [25]:
data.groupby("director_name").get_group("Adam McKay")

id_x  budget  popularity  revenue  \
176  43882   100.0          24   170.43   
323  44151    72.5          12   162.97   
366  44236    65.0          22   128.11   
505  44503    50.0          38   173.65   
839  45301    28.0          57   133.35   
916  45443    26.0          29    90.57   

                                           title  vote_average  vote_count  \
176                               The Other Guys           6.1        1383   
323  Talladega Nights: The Ballad of Ricky Bobby           6.2         491   
366                                Step Brothers           6.5        1062   
505            Anchorman 2: The Legend Continues           6.0         923   
839                                The Big Short           7.3        2607   
916        Anchorman: The Legend of Ron Burgundy           6.7        1493   

     director_id  year month        day director_name  gender  profit  \
176         4925  2010   Aug     Friday    Adam McKay       0   70.43   
323         4925  2006   Aug     Friday    Adam McKay       0   90.47   
366         4925  2008   Jul     Friday    Adam McKay       0   63.11   
505         4925  2013   Dec  Wednesday    Adam McKay       0  123.65   
839         4925  2015   Dec     Friday    Adam McKay       0  105.35   
916         4925  2004   Jul     Friday    Adam McKay       0   64.57   

     new_profit  
176       70.43  
323       90.47  
366       63.11  
505      123.65  
839      105.35  
916       64.57

* **data.groupby("director_name") gives us a group by director's name. Lets say we want to find how many movies have been directed by each director.**
* **So we write ["title"].count() after groupby. Our aggregation method here is count(). We can write any column name here instead of title, it will give the same answer.**
* **How to read this? -- For every group called director's name, count the number of rows or title values they have.** 
* In the output, we can see all 199 groups and next to them are total count of titles that have been directed by them. In other words, for every group, it will apply this aggregation called count. 

In [26]:
data.groupby("director_name")["title"].count()

director_name
Adam McKay                      6
Adam Shankman                   8
Alejandro González Iñárritu     6
Alex Proyas                     5
Alexander Payne                 5
                               ..
Wes Craven                     10
Wolfgang Petersen               7
Woody Allen                    18
Zack Snyder                     7
Zhang Yimou                     6
Name: title, Length: 199, dtype: int64

<hr style="border: 1px solid gray;">

**<code style="background:yellow;color:black">Quiz:</code> Given a dataframe of employee rating info (emp_id, monthly_rating). How will you calculate average rating of each employee?**

A. data.groupby('emp_id')

B. data.groupby('monthly_rating')

C. data.groupby('emp_id').mean()

D. data.groupby('monthly_rating').mean()

**Ans: C**

* Here, option B and D cannot be the answer because we have to groupby the column of emp_id because its asking How will you calculate average rating of each employee. In option A groupby is done but no aggregation is done. So option C we get groupby on emp_id and aggregation of average also. 

* **Here the syntax is tricky because in above code, data.groupby("director_name")["title"].count(), we have title as column name, whereas here we have data.groupby('emp_id').mean(), column name inside groupby only. In this syntax, it calculates mean() over every column where it is possible. Here, there are only 2 columns, so it will automatically apply mean over monthly_rating as well.**

<hr style="border: 1px solid gray;">

**<code style="background:yellow;color:black">Quiz:</code> For the following code, what will happen -- data.groupby('director_name').mean()**

A. It will give an error since we can't aggregate string columns like "title" using mean

B. It will show results only for columns where mean() aggregation is possible

C. It will show results for all the columns, with the non-applicable columns being converted to their integer/ascii format for computation

**Ans: B**

<hr style="border: 1px solid gray;">

# <code style="background:yellow;color:black">Multiple aggregations:</code>

**In pandas, when we perform groupby operations, we can apply <code style="background:yellow;color:black">multiple aggregation functions</code> to different columns using the agg() or aggregate() method. This allows us to compute various summary statistics for each group in a single step.**

### <code style="background:yellow;color:black">aggregate() or agg() method:</code>
    
* **In pandas, <code style="background:yellow;color:black">agg() and aggregate() methods</code> are essentially the same method with different names. They both serve the same purpose and can be used interchangeably. Both methods are used for aggregation operations in the context of a grouped DataFrame.**
* So, we can choose to use either agg() or aggregate() based on our preference.
* **The aggregate() method in pandas is used to apply one or more aggregation functions to the grouped data in a DataFrame. When we use groupby() to group a DataFrame based on one or more columns, we can use aggregate() to compute various summary statistics for each group.**

**<code style="background:yellow;color:black">Question:</code> For every director, find the first year and the last year in which they have directed a movie.**

* **Here we need to perform groupby on director's name and on year column we have to do two aggregations called min and max.** 
* When we are doing single aggregation we can simply write the aggregate method's name using the dot operator. 
* **But when we want to do multiple aggregations, we use this aggregate() or agg() method and then aggregate methods are passed in it as string parameters.**

In [27]:
data.groupby("director_name")["year"].aggregate(["min", "max"])

min   max
director_name                          
Adam McKay                   2004  2015
Adam Shankman                2001  2012
Alejandro González Iñárritu  2000  2015
Alex Proyas                  1994  2016
Alexander Payne              1999  2013
...                           ...   ...
Wes Craven                   1984  2011
Wolfgang Petersen            1981  2006
Woody Allen                  1977  2013
Zack Snyder                  2004  2016
Zhang Yimou                  2002  2014

[199 rows x 2 columns]

**<code style="background:yellow;color:black">Quiz:</code> For each year , how you will get the average, lowest, and highest value of ratings of movies?**

* **This is another way or syntax to write aggregation wherein we provide a dictionary.**
* This syntax means that over the vote_average column apply all the 3 aggregations.
* **This is an advanced syntax where we can perform <code style="background:yellow;color:black">multiple aggregates on multiple columns simultaneously</code> by providing another key:value pair in the .agg() method.**

In [28]:
data.groupby("year").agg(
    {
        "vote_average" : ["mean", "min", "max"]
    }
)

vote_average          
             mean  min  max
year                       
1976     7.633333  7.2  8.0
1977     6.900000  6.1  7.8
1978     7.020000  5.9  7.9
1979     7.025000  5.6  8.0
1980     7.187500  6.4  7.9
1981     6.927273  6.0  7.9
1982     6.810000  5.5  7.9
1983     6.666667  4.5  8.0
1984     6.760000  6.1  7.3
1985     6.885714  5.9  8.0
1986     6.809091  5.8  7.8
1987     6.991667  6.1  7.6
1988     6.692308  5.7  7.6
1989     6.750000  5.5  7.6
1990     6.661111  5.8  8.2
1991     7.100000  6.6  8.1
1992     6.640909  5.7  8.0
1993     7.144444  6.1  8.3
1994     6.596774  5.0  8.3
1995     6.509091  4.5  8.1
1996     6.384783  5.4  7.8
1997     6.552174  4.1  7.9
1998     6.442222  4.9  7.9
1999     6.384375  5.1  8.3
2000     6.347458  4.4  8.1
2001     6.217742  3.7  8.0
2002     6.359420  3.4  8.0
2003     6.115517  4.4  8.1
2004     6.427869  4.7  7.7
2005     6.229577  3.1  7.7
2006     6.212329  3.6  8.0
2007     6.307407  3.2  7.9
2008     6.268657  3.0  8.2
2009     6.190909  3.9  7.9
2010     6.223810  4.2  8.1
2011     6.098507  4.1  7.7
2012     6.216327  4.9  7.8
2013     6.315789  3.3  7.9
2014     6.313208  4.1  8.1
2015     6.141463  3.9  7.7
2016     5.915385  4.5  7.1

<hr style="border: 1px solid gray;">

**<code style="background:yellow;color:black">Question:</code> Give the details of the movies by high budget directors.**

* Whats the definition of high budget directors? Lets assume, high budget director is any director with atleast one movie with budget > 100 million.
* Can we reframe the question as -- Give the details of directors who made atleast one movie with budget > 100 million. 
<br><br>
* Here, .reset_index() method is used to get a seperate index values and both the columns of director_name and budget as well, otherwise we would not have got index values as a seperate column.
* **We got this. Now how can we filter out those directors who have done atleast one movie more than 100 million budget?** 
* **<code style="background:yellow;color:black">For this, we can do masking.</code> We take dataframe with max values i.e. data_dir_budget, apply explicit index loc to access budget column and apply mask to it and save or store it in a new df called "names".**
* When we check names, we have got all the names of those directors who are fitting into this criteria -- data_dir_budget["budget"] > 100.
<br><br>
* **Now, <code style="background:yellow;color:black">from the original dataframe we need to filter out those rows</code> where director is any of the ones present in "names" and our question is solved.**
* We can write .isin() if we want to have any one of the directors.
* There are 636 such rows where director is any of them.

In [30]:
data_dir_budget = data.groupby("director_name")["budget"].max().reset_index()

data_dir_budget

director_name  budget
0                     Adam McKay   100.0
1                  Adam Shankman    80.0
2    Alejandro González Iñárritu   135.0
3                    Alex Proyas   140.0
4                Alexander Payne    30.0
..                           ...     ...
194                   Wes Craven    40.0
195            Wolfgang Petersen   175.0
196                  Woody Allen    30.0
197                  Zack Snyder   250.0
198                  Zhang Yimou    94.0

[199 rows x 2 columns]

In [51]:
names = data_dir_budget.loc[data_dir_budget["budget"] > 100, "director_name"]

names

2      Alejandro González Iñárritu
3                      Alex Proyas
5                   Andrew Adamson
10                         Ang Lee
15                Barry Sonnenfeld
                  ...             
185                      Tim Story
187                    Tom Shadyac
188                     Tom Tykwer
195              Wolfgang Petersen
197                    Zack Snyder
Name: director_name, Length: 79, dtype: object

In [52]:
data.loc[data["director_name"].isin(names)]

id_x  budget  popularity  revenue  \
0     43597  237.00         150  2787.97   
1     43598  300.00         139   961.00   
2     43599  245.00         107   880.67   
3     43600  250.00         112  1084.94   
4     43602  258.00         115   890.87   
...     ...     ...         ...      ...   
1450  48267    0.40          33   100.00   
1451  48268    0.20          13     4.51   
1452  48274    0.00           5     2.61   
1458  48335    0.06          27     3.22   
1460  48363    0.00           3     0.32   

                                         title  vote_average  vote_count  \
0                                       Avatar           7.2       11800   
1     Pirates of the Caribbean: At World's End           6.9        4500   
2                                      Spectre           6.3        4466   
3                        The Dark Knight Rises           7.6        9106   
4                                 Spider-Man 3           5.9        3576   
...                                        ...           ...         ...   
1450                                   Mad Max           6.6        1213   
1451                                  Swingers           6.8         253   
1452                                     Three           6.3          31   
1458                                        Pi           7.1         586   
1460                            The Last Waltz           7.9          64   

      director_id  year month       day      director_name  gender   profit  \
0            4762  2009   Dec  Thursday      James Cameron       0  2550.97   
1            4763  2007   May  Saturday     Gore Verbinski       0   661.00   
2            4764  2015   Oct    Monday         Sam Mendes       0   635.67   
3            4765  2012   Jul    Monday  Christopher Nolan       0   834.94   
4            4767  2007   May   Tuesday          Sam Raimi       0   632.87   
...           ...   ...   ...       ...                ...     ...      ...   
1450         4845  1979   Apr  Thursday      George Miller       0    99.60   
1451         4813  1996   Oct    Friday         Doug Liman       0     4.31   
1452         4936  2010   Dec  Thursday         Tom Tykwer       0     2.61   
1458         4881  1998   Jul    Friday   Darren Aronofsky       0     3.16   
1460         4809  1978   May    Monday    Martin Scorsese       0     0.32   

      new_profit  
0        2550.97  
1         661.00  
2         635.67  
3         834.94  
4         632.87  
...          ...  
1450       99.60  
1451        4.31  
1452        2.61  
1458        3.16  
1460        0.32  

[636 rows x 15 columns]

# <code style="background:yellow;color:black">Groupby filtering:</code>

* **<code style="background:yellow;color:black">Groupby filtering</code> in pandas involves applying a condition to groups of data and filtering out groups that do not satisfy the condition.**
* **We can use the groupby() method along with filtering method like filter() to achieve this.**
***
* Is the above question same as filtering those rows where budget is greater than 100 million? So if we see this question -- Display only those rows where budget > 100 M. Is this question same as the question we have solved just now? -- NO.
* **This question simply asks show me those rows where budget is more than 100 million. Whereas the above question asks if a director has done any one movie greater than 100 million, show me all their movies.** If this particular director James Cameroon for example, has done even one movie greater than 100 million, we will show all of his movies irrespective of their budgets.
* **This is called <code style="background:yellow;color:black">Groupby filtering.</code>**
* **Do we agree that either we are selecting all the rows of that director or none of the rows of that director? -- YES. Either select the entire group either reject the entire group.**
* **This concept is very similar to <code style="background:yellow;color:black">"groupby and having in sql".</code> Having will either select the entire group based on condition or will reject the entire group**
***
* **We have achieved this groupby filtering through various steps as above. But we can use one line method called filter() on top of groupby() to achieve groupby filtering with just one line of code. Earlier we were applying a function, this time we will filter the groups.**

### <code style="background:yellow;color:black">filter() method:</code>

* **In pandas, the <code style="background:yellow;color:black">filter() method</code> is used to subset a DataFrame based on a particular condition. The filter() method can be applied to both rows and columns, and its behavior depends on the type of criteria specified.**
* When used with groupby objects, the filter() method is particularly useful. It allows us to filter groups of data based on group-specific conditions.
***
* **Now in the code cell below, firstly data.groupby("director_name") will groupby director name.**
* **Now over that group, we are applying a filtering criteria.** 
* **What happens is that each group is treated like one dataframe, we are picking that dataframe and applying max() over it on the budget column, then we are checking is it greater than 100 or not.**
* **If this condition is true, we will filter in that group, if this condition is false we will filter out that group.**
* **Meaning is- either we are selecting the entire group based on condition or either we are rejecting the entire group. So, if max budget of that group is greater than 100, select that group, else reject that group.**
* Result is same as achieved above through various steps as discussed.
***
Very very important point is that **<code style="background:yellow;color:black">in loc we are filtering rows, and in filter() we are filtering groups.</code>**

In [53]:
data.groupby("director_name").filter(lambda df: df["budget"].max() > 100)

id_x  budget  popularity  revenue  \
0     43597  237.00         150  2787.97   
1     43598  300.00         139   961.00   
2     43599  245.00         107   880.67   
3     43600  250.00         112  1084.94   
4     43602  258.00         115   890.87   
...     ...     ...         ...      ...   
1450  48267    0.40          33   100.00   
1451  48268    0.20          13     4.51   
1452  48274    0.00           5     2.61   
1458  48335    0.06          27     3.22   
1460  48363    0.00           3     0.32   

                                         title  vote_average  vote_count  \
0                                       Avatar           7.2       11800   
1     Pirates of the Caribbean: At World's End           6.9        4500   
2                                      Spectre           6.3        4466   
3                        The Dark Knight Rises           7.6        9106   
4                                 Spider-Man 3           5.9        3576   
...                                        ...           ...         ...   
1450                                   Mad Max           6.6        1213   
1451                                  Swingers           6.8         253   
1452                                     Three           6.3          31   
1458                                        Pi           7.1         586   
1460                            The Last Waltz           7.9          64   

      director_id  year month       day      director_name  gender   profit  \
0            4762  2009   Dec  Thursday      James Cameron       0  2550.97   
1            4763  2007   May  Saturday     Gore Verbinski       0   661.00   
2            4764  2015   Oct    Monday         Sam Mendes       0   635.67   
3            4765  2012   Jul    Monday  Christopher Nolan       0   834.94   
4            4767  2007   May   Tuesday          Sam Raimi       0   632.87   
...           ...   ...   ...       ...                ...     ...      ...   
1450         4845  1979   Apr  Thursday      George Miller       0    99.60   
1451         4813  1996   Oct    Friday         Doug Liman       0     4.31   
1452         4936  2010   Dec  Thursday         Tom Tykwer       0     2.61   
1458         4881  1998   Jul    Friday   Darren Aronofsky       0     3.16   
1460         4809  1978   May    Monday    Martin Scorsese       0     0.32   

      new_profit  
0        2550.97  
1         661.00  
2         635.67  
3         834.94  
4         632.87  
...          ...  
1450       99.60  
1451        4.31  
1452        2.61  
1458        3.16  
1460        0.32  

[636 rows x 15 columns]

<hr style="border: 1px solid gray;">

**<code style="background:yellow;color:black">Question: Categorize a movie as risky or non risky.</code> A movie is risky if its budget is greater than the average revenue of all the movies of that director.**

* **Lets say a director has done 10 movies. For every movie, we have to tell whether that movie is risky or not.** 
* **<code style="background:yellow;color:black">So, we have to take average revenue of these 10 movies and we have to compare with budget of one movie.</code>**
* **And if the budget this one movie is greater than the average revenue of this director then its risky, if no then its not risky.**

**Which concept in sql is this similar to? -- <code style="background:yellow;color:black">WINDOW FUNCTION.</code> For every group, we want aggregated value of that row. We take the average of the entire window and then we use to compare aggregate for each row.**

* In below code cell, basically what has happened is, df["revenue"].mean() this aggregation is being performed on the revenue column of that group because first we are doing .groupby(). 
* So specifically, within that group, this average is being computed for each row.

In [59]:
def check_risky(df):
    risky_flag = df["budget"] > df["revenue"].mean()
    df["risky"] = risky_flag
    return df

In [60]:
result = data.groupby("director_name", group_keys = True).apply(check_risky)

result

id_x  budget  popularity  revenue  \
director_name                                            
Adam McKay    176   43882   100.0          24   170.43   
              323   44151    72.5          12   162.97   
              366   44236    65.0          22   128.11   
              505   44503    50.0          38   173.65   
              839   45301    28.0          57   133.35   
...                   ...     ...         ...      ...   
Zhang Yimou   590   44692     0.0           9     0.00   
              604   44733    31.0          23   177.39   
              1217  46460     0.0          21    92.86   
              1223  46493     0.0           1     0.00   
              1389  47489     0.0           6     0.00   

                                                          title  vote_average  \
director_name                                                                   
Adam McKay    176                                The Other Guys           6.1   
              323   Talladega Nights: The Ballad of Ricky Bobby           6.2   
              366                                 Step Brothers           6.5   
              505             Anchorman 2: The Legend Continues           6.0   
              839                                 The Big Short           7.3   
...                                                         ...           ...   
Zhang Yimou   590                    Curse of the Golden Flower           6.6   
              604                                          Hero           7.2   
              1217                      House of Flying Daggers           7.1   
              1223             A Woman, a Gun and a Noodle Shop           4.8   
              1389                                  Coming Home           6.9   

                    vote_count  director_id  year month        day  \
director_name                                                        
Adam McKay    176         1383         4925  2010   Aug     Friday   
              323          491         4925  2006   Aug     Friday   
              366         1062         4925  2008   Jul     Friday   
              505          923         4925  2013   Dec  Wednesday   
              839         2607         4925  2015   Dec     Friday   
...                        ...          ...   ...   ...        ...   
Zhang Yimou   590          203         4945  2006   Dec   Thursday   
              604          635         4945  2002   Dec   Thursday   
              1217         439         4945  2004   May  Wednesday   
              1223          13         4945  2009   Dec     Friday   
              1389          49         4945  2014   May     Friday   

                   director_name  gender  profit  new_profit  risky  
director_name                                                        
Adam McKay    176     Adam McKay       0   70.43       70.43  False  
              323     Adam McKay       0   90.47       90.47  False  
              366     Adam McKay       0   63.11       63.11  False  
              505     Adam McKay       0  123.65      123.65  False  
              839     Adam McKay       0  105.35      105.35  False  
...                          ...     ...     ...         ...    ...  
Zhang Yimou   590    Zhang Yimou       0    0.00        0.00  False  
              604    Zhang Yimou       0  146.39      146.39  False  
              1217   Zhang Yimou       0   92.86       92.86  False  
              1223   Zhang Yimou       0    0.00        0.00  False  
              1389   Zhang Yimou       0    0.00        0.00  False  

[1465 rows x 16 columns]

**<code style="background:yellow;color:black">Approach 2: More simpler way</code>**

1. for every director - find the avg revenue
2. merge this with the original dataframe
3. row wise comparison

This is implemented below:

In [57]:
# Calculate average revenue for each director
avg_revenue_by_director = data.groupby('director_name')['revenue'].mean().reset_index()
avg_revenue_by_director.rename(columns = {'revenue': 'avg_revenue'}, inplace = True)

# Merge with the original dataframe
data = pd.merge(data, avg_revenue_by_director, on = 'director_name', how = 'left')

# Create a 'risky' column based on the comparison
data['risky'] = data['budget'] > data['avg_revenue']

# Drop the 'avg_revenue' column if you don't need it in the final result
data.drop('avg_revenue', axis = 1, inplace = True)

data

id_x  budget  popularity  revenue  \
0     43597  237.00         150  2787.97   
1     43598  300.00         139   961.00   
2     43599  245.00         107   880.67   
3     43600  250.00         112  1084.94   
4     43602  258.00         115   890.87   
...     ...     ...         ...      ...   
1460  48363    0.00           3     0.32   
1461  48370    0.03          19     3.15   
1462  48375    0.00           7     0.00   
1463  48376    0.00           3     0.00   
1464  48395    0.22          14     2.04   

                                         title  vote_average  vote_count  \
0                                       Avatar           7.2       11800   
1     Pirates of the Caribbean: At World's End           6.9        4500   
2                                      Spectre           6.3        4466   
3                        The Dark Knight Rises           7.6        9106   
4                                 Spider-Man 3           5.9        3576   
...                                        ...           ...         ...   
1460                            The Last Waltz           7.9          64   
1461                                    Clerks           7.4         755   
1462                                   Rampage           6.0         131   
1463                                   Slacker           6.4          77   
1464                               El Mariachi           6.6         238   

      director_id  year month       day      director_name  gender   profit  \
0            4762  2009   Dec  Thursday      James Cameron       0  2550.97   
1            4763  2007   May  Saturday     Gore Verbinski       0   661.00   
2            4764  2015   Oct    Monday         Sam Mendes       0   635.67   
3            4765  2012   Jul    Monday  Christopher Nolan       0   834.94   
4            4767  2007   May   Tuesday          Sam Raimi       0   632.87   
...           ...   ...   ...       ...                ...     ...      ...   
1460         4809  1978   May    Monday    Martin Scorsese       0     0.32   
1461         5369  1994   Sep   Tuesday        Kevin Smith       0     3.12   
1462         5148  2009   Aug    Friday           Uwe Boll       0     0.00   
1463         5535  1990   Jul    Friday  Richard Linklater       0     0.00   
1464         5097  1992   Sep    Friday   Robert Rodriguez       1     1.82   

      new_profit  risky  
0        2550.97  False  
1         661.00  False  
2         635.67  False  
3         834.94  False  
4         632.87  False  
...          ...    ...  
1460        0.32  False  
1461        3.12  False  
1462        0.00  False  
1463        0.00  False  
1464        1.82  False  

[1465 rows x 16 columns]

* **Now we have created a filter. Now how can we give all the risky movies? By the following code-**
* **So we can see that, there are 131 movies whose budget is more than the average revenue of that director. <code style="background:yellow;color:black">These movies are risky.</code>**
* Lets say Christopher Nolan has done 10 movies, and the average revenue that he has made is 100 million dollars. 
* Now Nolan is thinking of making a movie whose budget is 150 million dollars. So its somewhat risky; it wont necessarily do flop show, but it might be risky. 
* Because Nolan usually makes a revenue of 100 million, the budget of this new proposed movie is more than that average revenue.

In [58]:
result[result["risky"] == True]

id_x  budget  popularity  revenue  \
director_name                                                 
Andrzej Bartkowiak 349   44192    60.0          29    55.99   
Atom Egoyan        946   45538    25.0           4     0.00   
                   1194  46370    15.0          26     8.46   
                   1347  47224     5.0           7     3.26   
Brett Ratner       24    43630   210.0           3   459.36   
...                        ...     ...         ...      ...   
Uwe Boll           944   45536    25.0           7     2.41   
                   1058  45834    20.0           9    10.44   
                   1383  47453     3.5           4     0.00   
Wayne Wang         468   44419    55.0          19   154.91   
Zhang Yimou        192   43914    94.0          12    95.31   

                                               title  vote_average  \
director_name                                                        
Andrzej Bartkowiak 349                          Doom           5.0   
Atom Egoyan        946          Where the Truth Lies           5.9   
                   1194                        Chloe           5.9   
                   1347          The Sweet Hereafter           6.8   
Brett Ratner       24          X-Men: The Last Stand           6.3   
...                                              ...           ...   
Uwe Boll           944                    BloodRayne           3.5   
                   1058            Alone in the Dark           3.1   
                   1383  In the Name of the King III           3.3   
Wayne Wang         468             Maid in Manhattan           5.6   
Zhang Yimou        192            The Flowers of War           7.1   

                         vote_count  director_id  year month        day  \
director_name                                                             
Andrzej Bartkowiak 349          609         5061  2005   Oct   Thursday   
Atom Egoyan        946           66         5599  2005   Oct     Friday   
                   1194         498         5599  2009   Mar  Wednesday   
                   1347         103         5599  1997   May  Wednesday   
Brett Ratner       24          3525         4786  2006   May  Wednesday   
...                             ...          ...   ...   ...        ...   
Uwe Boll           944          118         5148  2005   Oct   Saturday   
                   1058         173         5148  2005   Jan     Friday   
                   1383          19         5148  2013   Dec     Friday   
Wayne Wang         468          485         5162  2002   Dec     Friday   
Zhang Yimou        192          187         4945  2011   Dec   Thursday   

                              director_name  gender  profit  new_profit  risky  
director_name                                                                   
Andrzej Bartkowiak 349   Andrzej Bartkowiak       0   -4.01       -4.01   True  
Atom Egoyan        946          Atom Egoyan       0  -25.00      -25.00   True  
                   1194         Atom Egoyan       0   -6.54       -6.54   True  
                   1347         Atom Egoyan       0   -1.74       -1.74   True  
Brett Ratner       24          Brett Ratner       0  249.36      249.36   True  
...                                     ...     ...     ...         ...    ...  
Uwe Boll           944             Uwe Boll       0  -22.59      -22.59   True  
                   1058            Uwe Boll       0   -9.56       -9.56   True  
                   1383            Uwe Boll       0   -3.50       -3.50   True  
Wayne Wang         468           Wayne Wang       1   99.91       99.91   True  
Zhang Yimou        192          Zhang Yimou       0    1.31        1.31   True  

[131 rows x 16 columns]